File: internal/compiler/lint/rules/jsx-a11y/noAriaUnsupportedElements.ts
import {createVisitor, signals} from "@internal/compiler";
import {descriptions} from "@internal/diagnostics";
import {JSXElement} from "@internal/ast";
import {hasJSXAttribute, isJSXElement} from "@internal/js-ast-utils";
import {ARIAProperty, ariaPropsMap} from "@internal/compiler/lint/utils/aria";

function hasAriaAttributes(node: JSXElement): boolean {
	const hasRole = hasJSXAttribute(node, "role");

	return (
		hasRole ||
		node.attributes.some((attr) =>
			attr.type === "JSXAttribute" &&
			attr.name.type === "JSXIdentifier" &&
			ariaPropsMap.has((attr.name.name as ARIAProperty))
		)
	);
}

export default createVisitor({
	name: "jsx-a11y/noAriaUnsupportedElements",
	enter(path) {
		const {node} = path;

		if (
			!(isJSXElement(node, "meta") ||
			isJSXElement(node, "html") ||
			isJSXElement(node, "script") ||
			isJSXElement(node, "style"))
		) {
			return signals.retain;
		}

		if (hasAriaAttributes(node)) {
			path.context.addNodeDiagnostic(
				node,
				descriptions.LINT.JSX_A11Y_ARIA_UNSUPPORTED_ELEMENTS,
			);
		}

		return signals.retain;
	},
});

File: crates/rome_js_analyze/src/aria_analyzers/nursery/no_aria_unsupported_elements.rs
use crate::aria_services::Aria;
use rome_analyze::{context::RuleContext, declare_rule, Rule, RuleDiagnostic};
use rome_console::markup;
use rome_js_syntax::jsx_ext::AnyJsxElement;
use rome_rowan::{AstNode, AstNodeList};

declare_rule! {
    /// Enforce that elements that do not support ARIA roles, states, and properties do not have those attributes.
    ///
    /// Source: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/aria-unsupported-elements.md
    ///
    /// ## Examples
    ///
    /// ### Invalid
    ///
    /// ```jsx,expect_diagnostic
    /// <meta charset="UTF-8" role="meta" />
    /// ```
    ///
    /// ```jsx,expect_diagnostic
    /// <html aria-required="true" />
    /// ```
    ///
    /// ## Valid
    ///
    /// ```jsx
    /// <meta charset="UTF-8" />
    /// ```
    ///
    /// ```jsx
    /// <html></html>
    /// ```
    ///
    ///
    pub(crate) NoAriaUnsupportedElements {
        version: "next",
        name: "noAriaUnsupportedElements",
        recommended: true,
    }
}

const ARIA_UNSUPPORTED_ELEMENTS: [&str; 4] = ["meta", "html", "script", "style"];

#[derive(Debug)]
enum AttributeKind {
    Role,
    Aria,
}

impl AttributeKind {
    /// Converts an [AttributeKind] to a string.
    fn as_str(&self) -> &'static str {
        match self {
            AttributeKind::Role => "role",
            AttributeKind::Aria => "aria-*",
        }
    }
}

#[derive(Debug)]
pub struct RuleState {
    attribute_kind: AttributeKind,
}

impl Rule for NoAriaUnsupportedElements {
    type Query = Aria<AnyJsxElement>;
    type State = RuleState;
    type Signals = Option<Self::State>;
    type Options = ();

    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
        let node = ctx.query();
        let aria_properties = ctx.aria_properties();

        let element_name = node.name().ok()?.as_jsx_name()?.value_token().ok()?;
        let element_name = element_name.text_trimmed();

        if ARIA_UNSUPPORTED_ELEMENTS.contains(&element_name) {
            // Check if the unsupported element has `role` or `aria-*` attribute
            let report = node.attributes().iter().find_map(|attribute| {
                let attribute = attribute.as_jsx_attribute()?;
                let attribute_name = attribute.name().ok()?.as_jsx_name()?.value_token().ok()?;

                if attribute_name.text_trimmed().starts_with("aria-")
                    && aria_properties
                        .get_property(attribute_name.text_trimmed())
                        .is_some()
                {
                    return Some(RuleState {
                        attribute_kind: AttributeKind::Aria,
                    });
                }

                if attribute_name.text_trimmed() == "role" {
                    return Some(RuleState {
                        attribute_kind: AttributeKind::Role,
                    });
                }
                None
            });
            return report;
        }

        None
    }

    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
        let node = ctx.query();
        let attribute_kind = state.attribute_kind.as_str();
        Some(
            RuleDiagnostic::new(
                rule_category!(),
                node.range(),
                markup! {
                    "Avoid the "<Emphasis>"role"</Emphasis>" attribute and "<Emphasis>"aria-*"</Emphasis>" attributes when using "<Emphasis>"meta"</Emphasis>", "<Emphasis>"html"</Emphasis>", "<Emphasis>"script"</Emphasis>", and "<Emphasis>"style"</Emphasis>" elements."
                },
            )
            .note(markup! {
                "Using "{attribute_kind}" on elements that do not support them can cause issues with screen readers."
            }),
        )
    }
}
